sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -1055,43 +1059,89 @@ case class GpuCast(
 
   private def castDecimalToDecimal(input: ColumnVector,
       from: DecimalType,
-      to: DecimalType): ColumnVector = {
+      to: DecimalType,
+      ansiMode: Boolean): ColumnVector = {
 
-    // At first, we conduct overflow check onto input column.
-    // Then, we cast checked input into target decimal type.
-    val checkedInput = if (to.scale <= from.scale) {
-      // No need to promote precision unless target scale is larger than the source one,
-      // which indicates the cast is always valid when to.scale <= from.scale.
-      input.incRefCount()
-    } else {
+    val isFrom32Bit = DecimalType.is32BitDecimalType(from)
+    val isTo32Bit = DecimalType.is32BitDecimalType(to)
+    val cudfDecimal = DecimalUtil.createCudfDecimal(to.precision, to.scale)
+
+    def castCheckedDecimal(checkedInput: ColumnVector): ColumnVector = {
+      to.scale - from.scale match {
+        case 0 =>
+          if (isFrom32Bit == isTo32Bit) {
+            checkedInput.incRefCount()
+          } else {
+            // the input is already checked, just cast it
+            checkedInput.castTo(cudfDecimal)
+          }
+        case diff if diff > 0 =>
+          checkedInput.castTo(cudfDecimal)
+        case _ =>
+          withResource(checkedInput.round(to.scale, ai.rapids.cudf.RoundMode.HALF_UP)) {
+            rounded => rounded.castTo(cudfDecimal)
+          }
+      }
+    }
+
+    def checkForOverflow: ColumnVector = {
       // Check whether there exists overflow during promoting precision or not.
       // We do NOT use `Scalar.fromDecimal(-to.scale, math.pow(10, 18).toLong)` here, because
       // cuDF binaryOperation on decimal will rescale right input to fit the left one.
       // The rescaling may lead to overflow.
-      val absBound = math.pow(10, DType.DECIMAL64_MAX_PRECISION + from.scale - to.scale).toLong
-      if (ansiMode) {
+      // absBound is the maximum value that the input column can have before being casted
+      val prec = to.precision + from.scale - to.scale
```

Comment:
what is prec? it appears to be the intermediate precision that cudf will do internally.  If so prec is a fairly bad name for it.

Reply:
Its not the intermediate precision.  When moving from a smaller scale to a bigger scale (or 32-bit to 64-bit), the target type is able to hold much more values on the fraction side which leaves less room for the whole number. e.g.
```      
      999999.999 => from.scale = 3
      9999.99999 => to.scale = 5
```
this means that there could be values in the input that will not be able to fit in the target type, so we calculate the max number that should be in the input column before we can safely cast the values without overflowing. In the above example the source can have a maximum of 4 digits before decimal point and 3 after the decimal point. Therefore the values of absMax and absMin will be calculated based on the prec value.

Reply:
Then put that in the comments.